### PR TITLE
add Makefile

### PR DESCRIPTION
**Pull request by saeidakbari-ir**
Simple make file for ease of use
Discussion: #46

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	go get -u github.com/matryer/moq
+	# Reqire when adding proto and grpc compilation
+	# go get -u github.com/golang/protobuf/protoc-gen-go
+	# go get -u github.com/mwitkow/go-proto-validators/protoc-gen-govalidators
+	# go get -u github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway
+	# go get -u github.com/grpc-ecosystem/grpc-gateway/protoc-gen-swagger
+
+# format the go source files
+fmt:
+	go fmt ./...
```
Comment (DrmagicE) [inline Makefile:64]: `go fmt` is not necessary because `goimports` is a superset of `go fmt` and does everything `go fmt` do.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	goimports -w $(FILES)
```
Comment (DrmagicE) [inline Makefile:65]: prefer @goimports -w $(FILES), it is too noisy..

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+ # run go vet on the source files
+vet:
+	go vet ./...
+
+# generate godocs and start a local documentation webserver on port 8085
+doc:
+	godoc -http=:8085 -index
+
+# update golang dependencies
+update-dependencies:
```
Comment (DrmagicE) [inline Makefile:80]: Gmqtt is using gomodule for dependency management, not dep. Maybe we can remove this?

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+# generate all grpc files and mocks and build the go code
+build:
+	go build -o $(BUILD_DIR)/gmqtt ./cmd/gmqttd
+
+# generate mocks and run short tests
+test: generate-mocks
+	go test -v ./... -short
+
+# generate mocks and run all tests
+test-it: generate-mocks
```
Comment (DrmagicE) [inline Makefile:114]: I think make test is enough, maybe remove this?

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+run:
+	go run ./cmd/gmqttd start -c ./cmd/gmqttd/default_config.yml
+
+# generate all grpc files and mocks and build the go code
+build:
+	go build -o $(BUILD_DIR)/gmqtt ./cmd/gmqttd
+
+# generate mocks and run short tests
+test: generate-mocks
+	go test -v ./... -short
```
Comment (DrmagicE) [inline Makefile:111]: Needs to add race detection `-race` and removes `-short`:
go test -race ./...

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+# remove all generated artifacts and clean all build artifacts
+clean-all:
+	go clean -i ./...
+	rm -fr rpc
+
+# fetch and install all required tools
+tools:
+	go get -u golang.org/x/tools/cmd/goimports
+	go get -u github.com/golang/lint/golint
+	go get -u github.com/matryer/moq
```
Comment (DrmagicE) [inline Makefile:55]: The gomock library should be : `go get github.com/golang/mock/mockgen@v1.4.4`

File: Makefile (hunk 3 above)
Comment (saeidakbari-ir) [reply Makefile:80]: Sorry I miss that

**Pull request merged**